Show header when section item empty

marcuswu0814 opened the pull request:
## Changes in this pull request

Issue fixed: #1117

I adding a new constructor for making a `IGListCollectionViewLayout` instance that can always show sticky header although section data is empty.

It's working well and [this is demo project](https://github.com/marcuswu0814/IGListKit_ShowStickyHeaderWhenDataEmpty).

Bug I'm not sure is any good way to let this much more readability. Is there any good advice?

```objc
const CGRect headerBounds = (self.scrollDirection == UICollectionViewScrollDirectionVertical) ?
                CGRectMake(insets.left,
                        (itemCount == 0) ? CGRectGetMaxY(rollingSectionBounds) : CGRectGetMinY(rollingSectionBounds) - headerSize.height,
                        paddedLengthInFixedDirection,
                        headerSize.height) :
                CGRectMake((itemCount == 0) ? CGRectGetMaxX(rollingSectionBounds) : CGRectGetMinX(rollingSectionBounds) - headerSize.width,
                        insets.top,
                        headerSize.width,
                        paddedLengthInFixedDirection);
```



### Checklist

- [x] All tests pass. Demo project builds and runs.
- [x] I added tests, an experiment, or detailed why my change isn't tested.
- [x] I added an entry to the `CHANGELOG.md` for any breaking changes, enhancements, or bug fixes.
- [x] I have reviewed the [contributing guide](https://github.com/Instagram/IGListKit/blob/master/.github/CONTRIBUTING.md)

rnystrom commented on Source/IGListCollectionViewLayout.h:
```diff
@@ -128,6 +128,21 @@ NS_SWIFT_NAME(ListCollectionViewLayout)
+- (instancetype)initWithStickyHeaders:(BOOL)stickyHeaders
+                  showHeaderWhenEmpty:(BOOL)showHeaderWhenEmpty
+                      topContentInset:(CGFloat)topContentInset
+                        stretchToEdge:(BOOL)stretchToEdge;
```
Could we actually just include this as a mutable `@property BOOL showHeaderWhenEmpty`? Trying to prevent some initializer creep here, ha!

marcuswu0814 replied:
Sure, I will roll it back and extract the property `showHeaderWhenEmpty ` to public.

rnystrom commented on Tests/IGListCollectionViewLayoutTests.m:
```diff
@@ -86,6 +95,54 @@ - (void)test_whenEmptyData_thatContentSizeZero {
+    
+    IGAssertEqualFrame([self headerForSection:0].frame, 0, 0, 0, 0);
+    IGAssertEqualFrame([self headerForSection:1].frame, 0, 0, 0, 0);
+}
```
❤️ tests!

marcuswu0814 replied:
😄

rnystrom commented on Source/IGListCollectionViewLayout.mm:
```diff
@@ -577,19 +590,23 @@ - (void)calculateLayoutIfNeeded {
         _sectionData[section].headerBounds = headerBounds;
-
+        
+        if (itemCount == 0) {
```
Could we capture `itemCount == 0` in a variable?
```objc
const BOOL itemsEmpty = itemCount == 0;
```

marcuswu0814 replied:
Sure. 👍

rnystrom commented on Source/IGListCollectionViewLayout.mm:
```diff
@@ -577,19 +590,23 @@ - (void)calculateLayoutIfNeeded {
         const CGRect headerBounds = (self.scrollDirection == UICollectionViewScrollDirectionVertical) ?
                 CGRectMake(insets.left,
-                        CGRectGetMinY(rollingSectionBounds) - headerSize.height,
+                        (itemCount == 0) ? CGRectGetMaxY(rollingSectionBounds) : CGRectGetMinY(rollingSectionBounds) - headerSize.height,
```
What happens if `itemCount == 0` but `showHeaderWhenEmpty` is `NO`?

marcuswu0814 replied:
Nothing special, the `headerBounds` will same as `itemCount == 0`.

Before I added this condition, the `headerBounds`'s y position is out off screen when `itemCount == 0`. 

I think this changed did not affect any function because in method 

```objc
- (NSArray<UICollectionViewLayoutAttributes *> *)layoutAttributesForElementsInRect:(CGRect)rect {
        ...
        if (itemCount > 0 || self.showHeaderWhenEmpty) {
            for (NSString *elementKind in _supplementaryAttributesCache.allKeys) {
                NSIndexPath *indexPath = indexPathForSection(section);
                UICollectionViewLayoutAttributes *attributes = [self layoutAttributesForSupplementaryViewOfKind:elementKind
                                                                                                    atIndexPath:indexPath];
                // do not add zero height headers/footers or headers/footers that are outside the rect
                const CGRect frame = attributes.frame;
                const CGRect intersection = CGRectIntersection(frame, rect);
                if (!CGRectIsEmpty(intersection) && CGRectGetLengthInDirection(frame, self.scrollDirection) > 0.0) {
                    [result addObject:attributes];
                }
            }
        }
        ...
}
```

 the header will not be added into the result array.

So, should I modified this into `itemsEmpty && self.showHeaderWhenEmpty`? 😅)

marcuswu0814 replied:
Hum... I think I'm wrong, the UI will getting wrong layout. 

I will take a look later.
In my change and set `showHeaderWhenEmpty` to `NO`, try to set a section item count to zero and some error happed.

marcuswu0814 replied:
When I trying this, found another bug and note it below.

marcuswu0814 commented:
I think the bug is not cause by my change.

The following demo is using [this project](https://github.com/marcuswu0814/IGListKit_ShowStickyHeaderWhenDataEmpty).

### In branch `IG/release`
Before any thing change, used `release` version install by current pod latest version:
![ig release](https://user-images.githubusercontent.com/6396757/37762354-00eecfbe-2df7-11e8-83e9-e0f9fe987bbc.gif)

**Both header and footer had some bug.**

### In branch `IG/master`

![ig master](https://user-images.githubusercontent.com/6396757/37762389-1ea1cafc-2df7-11e8-890c-84e629eed456.gif)

**The footer work well**, maybe fixed it #1094, but header got some bug like release version.

### In my change, branch `marcuswu0814/feature`

![marcus0814 feature](https://user-images.githubusercontent.com/6396757/37762548-89b5d28e-2df7-11e8-8864-a7b6662d9ae9.gif)

If set property `showHeaderWhenEmpty` to `NO` and the header position will getting wrong.

So, In current `master` branch, if section item is empty and using sticky header, sticky header position will getting some bug.

marcuswu0814 commented:
I fixed this bug on my branch latest commit, and the old test case didn't cover this bug, I already modified the test case and pass it. Now, avoid this bug! 😎

![bugfix](https://user-images.githubusercontent.com/6396757/37770811-92bd1ea0-2e10-11e8-8c54-777b5e14d90e.gif)

Need open another issue to tracking this?
And if having any new issue, feel free to tell ms, thanks. 😅

marcuswu0814 commented:
Hi @rnystrom, will this possible be merged in 3.3.0 release?
